### PR TITLE
create table default to null

### LINKED ISSUE #5575: CREATE TABLE defaults to NON nullable (should be nullable to follow postgres)
`CREATE TABLE` statements create tables with non-nullable columns by default, which contradicts PostgreSQL 15.

```sql
CREATE TABLE table_without_values(field1 BIGINT, field2 BIGINT);
```
DDL query. It defaults to non-nullable columns while creating a table, which is not the case in [PostgreSQL 15](https://www.postgresql.org/docs/current/sql-createtable.html).

> NULL
> The column is allowed to contain null values. This is the default.
> This clause is only provided for compatibility with non-standard SQL databases. Its use is discouraged in new applications.

PS: The code piece that results non-nullable default:
https://github.com/apache/arrow-datafusion/blob/1a22f9fd436c9892566b668e535e0d6c8cb9fbd3/datafusion/sql/src/planner.rs#L127-L144

_Originally posted by @metesynnada in https://github.com/apache/arrow-datafusion/issues/5520#issuecomment-1465991022_
            

# Desired behavior:
Make the default columns NULLABLE 

this should work (and insert a single null value into `foo`)

```sql
create table foo(x int);
insert into foo values (null);
```

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #5575

# Rationale for this change
This part enhances the `CREATE` by default null table. 

# What changes are included in this PR?
```
create table foo(x int);
insert into foo values (null);
```
This statement will insert null into the table.
  
The `create.rs` module is no longer needed as its functionality has been merged into the planner. Additionally, the `run_query` function in `datafusion/core/tests/sqllogictests/src/engines/datafusion/mod.r`s no longer needs to check if the statement is a CREATE TABLE statement.

# Are these changes tested?
Add tests in `/datafusion/core/tests/sqllogictests/test_files/ddl.slt`

# Are there any user-facing changes?
Allow a traditional way to create a table and make it default to null
```
create table foo(x int);
```

### CODE REVIEW

File: datafusion/sql/src/statement.rs [hunk 1]
```diff
@@ -127,42 +129,66 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
             {
-                let plan = self.query_to_plan(*query, planner_context)?;
-                let input_schema = plan.schema();
-
-                let plan = if !columns.is_empty() {
-                    let schema = self.build_schema(columns)?.to_dfschema_ref()?;
-                    if schema.fields().len() != input_schema.fields().len() {
-                        return Err(DataFusionError::Plan(format!(
+                match query {
+                    Some(query) => {
```
Comment (comphead) [inline datafusion/sql/src/statement.rs:133]: looks like this code will handle traditoinal create table no just CTAS, do you think we can remove ```
            Statement::CreateTable { .. } => Err(DataFusionError::NotImplemented(
                "Only `CREATE TABLE table_name AS SELECT ...` statement is supported"
                    .to_string(),
            )),
            ```
            condition?

**Comment by metesynnada** [comment]
I will review it carefully tomorrow.

File: datafusion/core/tests/sqllogictests/test_files/ddl.slt [hunk 2]
```diff
@@ -553,3 +553,15 @@ set datafusion.sql_parser.parse_float_as_decimal = false;
...
+statement ok
+create table foo(x int);
+
+statement ok
+insert into foo values (null);
+
+query I
+select * from foo;
+----
+NULL
```
Comment (metesynnada) [inline datafusion/core/tests/sqllogictests/test_files/ddl.slt:567]: Cool 👍🏻

File: datafusion/sql/src/statement.rs [hunk 3]
```diff
@@ -127,42 +129,66 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                        let plan = self.query_to_plan(*query, planner_context)?;
+                        let input_schema = plan.schema();
+
+                        let plan = if !columns.is_empty() {
+                            let schema = self.build_schema(columns)?.to_dfschema_ref()?;
```
Comment (metesynnada) [inline datafusion/sql/src/statement.rs:138]: The problem of the `NOT NULL` default is still here since you did not change the `self.build_schema(columns)`.

This should be the default behaviour:
```rust
pub fn build_schema(&self, columns: Vec<SQLColumnDef>) -> Result<Schema> {
        let mut fields = Vec::with_capacity(columns.len());

        for column in columns {
            let data_type = self.convert_simple_data_type(&column.data_type)?;
            let prevent_null = column
                .options
                .iter()
                .any(|x| x.option == ColumnOption::NotNull);
            fields.push(Field::new(
                normalize_ident(column.name, self.options.enable_ident_normalization),
                data_type,
                !prevent_null,
            ));
        }

        Ok(Schema::new(fields))
    }
```
then, it would help if you changed the `sqllogictests`, for example, from
```
describe table_with_normalization
----
field1 Int64 NO
field2 Int64 NO
```
to
```
describe table_with_normalization
----
field1 Int64 YES
field2 Int64 YES
```

In this way, you can give the same defaults with PostgreSQL.

**Review by metesynnada — COMMENTED** [review]
Nice work, thanks 👍🏻 Although the implementation includes appending NULLs, it does not fully address the issue at hand. The root cause of the problem lies in the creation of a table with non-nullable columns as the default. Therefore, in order to resolve the issue completely, it is necessary to modify this behavior.

File: datafusion/sql/src/statement.rs [hunk 4]
```diff
@@ -127,42 +129,66 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-
-                Ok(LogicalPlan::CreateMemoryTable(CreateMemoryTable {
-                    name: self.object_name_to_table_reference(name)?,
-                    input: Arc::new(plan),
-                    if_not_exists,
-                    or_replace,
-                }))
+
+                    None => {
+                        let schema = self.build_schema(columns)?.to_dfschema_ref()?;
```
Comment (metesynnada) [inline datafusion/sql/src/statement.rs:175]: The problem of the `NOT NULL` default is still here since you did not change the `self.build_schema(columns)`.

This should be the default behaviour:
```rust
pub fn build_schema(&self, columns: Vec<SQLColumnDef>) -> Result<Schema> {
        let mut fields = Vec::with_capacity(columns.len());

        for column in columns {
            let data_type = self.convert_simple_data_type(&column.data_type)?;
            let prevent_null = column
                .options
                .iter()
                .any(|x| x.option == ColumnOption::NotNull);
            fields.push(Field::new(
                normalize_ident(column.name, self.options.enable_ident_normalization),
                data_type,
                !prevent_null,
            ));
        }

        Ok(Schema::new(fields))
    }
```
then, it would help if you changed the `sqllogictests`, for example, from
```
describe table_with_normalization
----
field1 Int64 NO
field2 Int64 NO
```
to
```
describe table_with_normalization
----
field1 Int64 YES
field2 Int64 YES
```

In this way, you can give the same defaults with PostgreSQL.

File: datafusion/sql/src/statement.rs [hunk 5]
```diff
@@ -127,42 +129,66 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                        let plan = EmptyRelation {
+                            produce_one_row: false,
+                            schema,
+                        };
+                        let plan = LogicalPlan::EmptyRelation(plan);
```
Comment (metesynnada) [inline datafusion/sql/src/statement.rs:180]: Cool idea 👍🏻

**Comment by metesynnada** [comment]
LGTM! Thanks for the effort.

File: datafusion/sql/src/planner.rs [hunk 6]
```diff
@@ -130,14 +130,14 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
         for column in columns {
             let data_type = self.convert_simple_data_type(&column.data_type)?;
-            let allow_null = column
+            let prevent_null = column
```
Comment (comphead) [inline datafusion/sql/src/planner.rs:133]: nit: I would name the variable nullable/not_nullable.

File: datafusion/core/tests/sqllogictests/src/engines/datafusion/mod.rs [hunk 7]
```diff
@@ -80,29 +77,8 @@ async fn run_query(ctx: &SessionContext, sql: impl Into<String>) -> Result<DFOut
...
-                    && table_properties.is_empty()
-                    && with_options.is_empty() =>
-                {
-                    return create_table(ctx, name, columns, if_not_exists, or_replace)
-                        .await
-                }
-                _ => {}
-            };
+        if let Statement::Statement(_) = statement0 {
+            {}
```
Comment (comphead) [inline datafusion/core/tests/sqllogictests/src/engines/datafusion/mod.rs:81]: nit: it might be good in such cases leave your own comments in the PR saying why this change was done. I noticed people here doing such way and that looks cool :)

**Comment by alamb** [comment]
Marking as draft as this PR has feedback pending. Let us know if you need help finishing this one up @Weijun-H

File: datafusion/sql/src/statement.rs [hunk 8]
```diff
@@ -181,10 +207,6 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     definition: sql,
                 }))
             }
-            Statement::CreateTable { .. } => Err(DataFusionError::NotImplemented(
```
Comment (alamb) [inline datafusion/sql/src/statement.rs:184]: ❤️

**Review by alamb — APPROVED** [review]
I tried this out locally with datafusion-cli and it works great 👍 

```
❯ create table foo (x int);
0 rows in set. Query took 0.003 seconds.
❯ insert into foo values (1);
0 rows in set. Query took 0.004 seconds.
❯ insert into foo values (2);
0 rows in set. Query took 0.002 seconds.
❯ insert into foo select * from foo;
0 rows in set. Query took 0.003 seconds.
❯ select * from foo;
+---+
| x |
+---+
| 1 |
| 2 |
| 1 |
| 2 |
+---+
4 rows in set. Query took 0.002 seconds.
```


```
❯ select * from foo;
+---+
| x |
+---+
| 1 |
| 2 |
| 1 |
| 2 |
|   |
+---+
5 rows in set. Query took 0.002 seconds.
```

Thank you @Weijun-H  and @comphead and @metesynnada for the reviews

**Comment by metesynnada** [comment]
Such a good PR, thanks for the effort @Weijun-H 👍🏻
